compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1407,7 +1417,11 @@ impl<'a> Parser<'a> {
         vis: Visibility,
         attrs: Vec<Attribute>,
     ) -> PResult<'a, FieldDef> {
-        let name = self.parse_field_ident(adt_ty, lo)?;
+        let name = self.parse_field_ident_or_underscore(adt_ty, lo)?;
+        let span = lo.to(self.prev_token.span);
+        if name.name == kw::Underscore {
+            self.sess.gated_spans.gate(sym::unnamed_fields, span);
+        }
```

Comment:
I think this logic should be moved to `fn parse_field_ident` making `fn parse_field_ident_or_underscore` unnecessary.
It's better to keep the same grammar for all fields including variant fields, underscore fields in variants will be rejected semantically in AST validation anyway.